// @remnawave/xtls-sdk/build/src/common/utils/decode-user/constants/account-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACCOUNT_TYPES = void 0;
const config_1 = require("../../../../xray-protos/proxy/shadowsocks_2022/config");
const config_2 = require("../../../../xray-protos/proxy/shadowsocks/config");
const config_3 = require("../../../../xray-protos/proxy/trojan/config");
const account_1 = require("../../../../xray-protos/proxy/vless/account");
const config_4 = require("../../../../xray-protos/proxy/socks/config");
const config_5 = require("../../../../xray-protos/proxy/http/config");
exports.ACCOUNT_TYPES = {
    [config_3.Account.$type]: {
        decoder: config_3.Account,
        protocol: 'trojan',
        getCredentials: (acc) => ({
            password: acc.password,
        }),
    },
    [account_1.Account.$type]: {
        decoder: account_1.Account,
        protocol: 'vless',
        getCredentials: (acc) => ({
            id: acc.id,
            flow: acc.flow,
            encryption: acc.encryption,
            xorMode: acc.xorMode,
            seconds: acc.seconds,
            padding: acc.padding,
            reverse: acc.reverse,
            testpre: acc.testpre,
            testseed: acc.testseed,
        }),
    },
    [config_2.Account.$type]: {
        decoder: config_2.Account,
        protocol: 'shadowsocks',
        getCredentials: (acc) => ({
            password: acc.password,
            cipherType: acc.cipherType,
            ivCheck: acc.ivCheck,
        }),
    },
    [config_1.Account.$type]: {
        decoder: config_1.Account,
        protocol: 'shadowsocks2022',
        getCredentials: (acc) => ({
            key: acc.key,
        }),
    },
    [config_4.Account.$type]: {
        decoder: config_4.Account,
        protocol: 'socks',
        getCredentials: (acc) => ({
            username: acc.username,
            password: acc.password,
        }),
    },
    [config_5.Account.$type]: {
        decoder: config_5.Account,
        protocol: 'http',
        getCredentials: (acc) => ({
            username: acc.username,
            password: acc.password,
        }),
    },
};
