// @remnawave/xtls-sdk/build/src/xray-protos/app/stats/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/stats/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelConfig = exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.stats";
function createBaseConfig() {
    return { $type: "xray.app.stats.Config" };
}
exports.Config = {
    $type: "xray.app.stats.Config",
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return { $type: exports.Config.$type };
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseConfig();
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function createBaseChannelConfig() {
    return { $type: "xray.app.stats.ChannelConfig", Blocking: false, SubscriberLimit: 0, BufferSize: 0 };
}
exports.ChannelConfig = {
    $type: "xray.app.stats.ChannelConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.Blocking !== false) {
            writer.uint32(8).bool(message.Blocking);
        }
        if (message.SubscriberLimit !== 0) {
            writer.uint32(16).int32(message.SubscriberLimit);
        }
        if (message.BufferSize !== 0) {
            writer.uint32(24).int32(message.BufferSize);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseChannelConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.Blocking = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.SubscriberLimit = reader.int32();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.BufferSize = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ChannelConfig.$type,
            Blocking: isSet(object.Blocking) ? globalThis.Boolean(object.Blocking) : false,
            SubscriberLimit: isSet(object.SubscriberLimit) ? globalThis.Number(object.SubscriberLimit) : 0,
            BufferSize: isSet(object.BufferSize) ? globalThis.Number(object.BufferSize) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.Blocking !== false) {
            obj.Blocking = message.Blocking;
        }
        if (message.SubscriberLimit !== 0) {
            obj.SubscriberLimit = Math.round(message.SubscriberLimit);
        }
        if (message.BufferSize !== 0) {
            obj.BufferSize = Math.round(message.BufferSize);
        }
        return obj;
    },
    create(base) {
        return exports.ChannelConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseChannelConfig();
        message.Blocking = object.Blocking ?? false;
        message.SubscriberLimit = object.SubscriberLimit ?? 0;
        message.BufferSize = object.BufferSize ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ChannelConfig.$type, exports.ChannelConfig);
function isSet(value) {
    return value !== null && value !== undefined;
}
