// @remnawave/xtls-sdk/build/src/stats/models/get-sys-stats/get-sys-stats.response.model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetSysStatsResponseModel = void 0;
const long_1 = __importDefault(require("long"));
class GetSysStatsResponseModel {
    constructor(data) {
        this.numGoroutine = long_1.default.fromValue(data.NumGoroutine).toNumber();
        this.numGC = long_1.default.fromValue(data.NumGC).toNumber();
        this.alloc = long_1.default.fromValue(data.Alloc).toNumber();
        this.totalAlloc = long_1.default.fromValue(data.TotalAlloc).toNumber();
        this.sys = long_1.default.fromValue(data.Sys).toNumber();
        this.mallocs = long_1.default.fromValue(data.Mallocs).toNumber();
        this.frees = long_1.default.fromValue(data.Frees).toNumber();
        this.liveObjects = long_1.default.fromValue(data.LiveObjects).toNumber();
        this.pauseTotalNs = long_1.default.fromValue(data.PauseTotalNs).toNumber();
        this.uptime = long_1.default.fromValue(data.Uptime).toNumber();
    }
}
exports.GetSysStatsResponseModel = GetSysStatsResponseModel;
