// @remnawave/xtls-sdk/build/src/xray-protos/common/net/address.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/net/address.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.IPOrDomain = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.common.net";
function createBaseIPOrDomain() {
    return { $type: "xray.common.net.IPOrDomain", ip: undefined, domain: undefined };
}
exports.IPOrDomain = {
    $type: "xray.common.net.IPOrDomain",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.ip !== undefined) {
            writer.uint32(10).bytes(message.ip);
        }
        if (message.domain !== undefined) {
            writer.uint32(18).string(message.domain);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseIPOrDomain();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.ip = reader.bytes();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.domain = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.IPOrDomain.$type,
            ip: isSet(object.ip) ? bytesFromBase64(object.ip) : undefined,
            domain: isSet(object.domain) ? globalThis.String(object.domain) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.ip !== undefined) {
            obj.ip = base64FromBytes(message.ip);
        }
        if (message.domain !== undefined) {
            obj.domain = message.domain;
        }
        return obj;
    },
    create(base) {
        return exports.IPOrDomain.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseIPOrDomain();
        message.ip = object.ip ?? undefined;
        message.domain = object.domain ?? undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.IPOrDomain.$type, exports.IPOrDomain);
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
