// @remnawave/xtls-sdk/build/src/xray-protos/common/net/destination.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/net/destination.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoint = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
const address_1 = require("./address");
const network_1 = require("./network");
exports.protobufPackage = "xray.common.net";
function createBaseEndpoint() {
    return { $type: "xray.common.net.Endpoint", network: 0, address: undefined, port: 0 };
}
exports.Endpoint = {
    $type: "xray.common.net.Endpoint",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.network !== 0) {
            writer.uint32(8).int32(message.network);
        }
        if (message.address !== undefined) {
            address_1.IPOrDomain.encode(message.address, writer.uint32(18).fork()).join();
        }
        if (message.port !== 0) {
            writer.uint32(24).uint32(message.port);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEndpoint();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.network = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.address = address_1.IPOrDomain.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.port = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Endpoint.$type,
            network: isSet(object.network) ? (0, network_1.networkFromJSON)(object.network) : 0,
            address: isSet(object.address) ? address_1.IPOrDomain.fromJSON(object.address) : undefined,
            port: isSet(object.port) ? globalThis.Number(object.port) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.network !== 0) {
            obj.network = (0, network_1.networkToJSON)(message.network);
        }
        if (message.address !== undefined) {
            obj.address = address_1.IPOrDomain.toJSON(message.address);
        }
        if (message.port !== 0) {
            obj.port = Math.round(message.port);
        }
        return obj;
    },
    create(base) {
        return exports.Endpoint.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseEndpoint();
        message.network = object.network ?? 0;
        message.address = (object.address !== undefined && object.address !== null)
            ? address_1.IPOrDomain.fromPartial(object.address)
            : undefined;
        message.port = object.port ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Endpoint.$type, exports.Endpoint);
function isSet(value) {
    return value !== null && value !== undefined;
}
