// @remnawave/xtls-sdk/build/src/xray-protos/common/net/network.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/net/network.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkList = exports.Network = exports.protobufPackage = void 0;
exports.networkFromJSON = networkFromJSON;
exports.networkToJSON = networkToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.common.net";
var Network;
(function (Network) {
    Network[Network["Unknown"] = 0] = "Unknown";
    Network[Network["TCP"] = 2] = "TCP";
    Network[Network["UDP"] = 3] = "UDP";
    Network[Network["UNIX"] = 4] = "UNIX";
    Network[Network["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Network || (exports.Network = Network = {}));
function networkFromJSON(object) {
    switch (object) {
        case 0:
        case "Unknown":
            return Network.Unknown;
        case 2:
        case "TCP":
            return Network.TCP;
        case 3:
        case "UDP":
            return Network.UDP;
        case 4:
        case "UNIX":
            return Network.UNIX;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Network.UNRECOGNIZED;
    }
}
function networkToJSON(object) {
    switch (object) {
        case Network.Unknown:
            return "Unknown";
        case Network.TCP:
            return "TCP";
        case Network.UDP:
            return "UDP";
        case Network.UNIX:
            return "UNIX";
        case Network.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseNetworkList() {
    return { $type: "xray.common.net.NetworkList", network: [] };
}
exports.NetworkList = {
    $type: "xray.common.net.NetworkList",
    encode(message, writer = new wire_1.BinaryWriter()) {
        writer.uint32(10).fork();
        for (const v of message.network) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseNetworkList();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag === 8) {
                        message.network.push(reader.int32());
                        continue;
                    }
                    if (tag === 10) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.network.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.NetworkList.$type,
            network: globalThis.Array.isArray(object?.network) ? object.network.map((e) => networkFromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.network?.length) {
            obj.network = message.network.map((e) => networkToJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.NetworkList.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseNetworkList();
        message.network = object.network?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.NetworkList.$type, exports.NetworkList);
