// @remnawave/xtls-sdk/build/src/xray-protos/common/net/port.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/net/port.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortList = exports.PortRange = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.common.net";
function createBasePortRange() {
    return { $type: "xray.common.net.PortRange", From: 0, To: 0 };
}
exports.PortRange = {
    $type: "xray.common.net.PortRange",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.From !== 0) {
            writer.uint32(8).uint32(message.From);
        }
        if (message.To !== 0) {
            writer.uint32(16).uint32(message.To);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePortRange();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.From = reader.uint32();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.To = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.PortRange.$type,
            From: isSet(object.From) ? globalThis.Number(object.From) : 0,
            To: isSet(object.To) ? globalThis.Number(object.To) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.From !== 0) {
            obj.From = Math.round(message.From);
        }
        if (message.To !== 0) {
            obj.To = Math.round(message.To);
        }
        return obj;
    },
    create(base) {
        return exports.PortRange.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePortRange();
        message.From = object.From ?? 0;
        message.To = object.To ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.PortRange.$type, exports.PortRange);
function createBasePortList() {
    return { $type: "xray.common.net.PortList", range: [] };
}
exports.PortList = {
    $type: "xray.common.net.PortList",
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.range) {
            exports.PortRange.encode(v, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePortList();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.range.push(exports.PortRange.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.PortList.$type,
            range: globalThis.Array.isArray(object?.range) ? object.range.map((e) => exports.PortRange.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.range?.length) {
            obj.range = message.range.map((e) => exports.PortRange.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.PortList.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePortList();
        message.range = object.range?.map((e) => exports.PortRange.fromPartial(e)) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.PortList.$type, exports.PortList);
function isSet(value) {
    return value !== null && value !== undefined;
}
