// @remnawave/xtls-sdk/build/src/xray-protos/app/version/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/version/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.version";
function createBaseConfig() {
    return { $type: "xray.app.version.Config", coreVersion: "", minVersion: "", maxVersion: "" };
}
exports.Config = {
    $type: "xray.app.version.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.coreVersion !== "") {
            writer.uint32(10).string(message.coreVersion);
        }
        if (message.minVersion !== "") {
            writer.uint32(18).string(message.minVersion);
        }
        if (message.maxVersion !== "") {
            writer.uint32(26).string(message.maxVersion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.coreVersion = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.minVersion = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.maxVersion = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            coreVersion: isSet(object.coreVersion) ? globalThis.String(object.coreVersion) : "",
            minVersion: isSet(object.minVersion) ? globalThis.String(object.minVersion) : "",
            maxVersion: isSet(object.maxVersion) ? globalThis.String(object.maxVersion) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.coreVersion !== "") {
            obj.coreVersion = message.coreVersion;
        }
        if (message.minVersion !== "") {
            obj.minVersion = message.minVersion;
        }
        if (message.maxVersion !== "") {
            obj.maxVersion = message.maxVersion;
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.coreVersion = object.coreVersion ?? "";
        message.minVersion = object.minVersion ?? "";
        message.maxVersion = object.maxVersion ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function isSet(value) {
    return value !== null && value !== undefined;
}
