// @remnawave/xtls-sdk/build/src/xray-protos/proxy/vmess/inbound/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/vmess/inbound/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.DefaultConfig = exports.DetourConfig = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const user_1 = require("../../../common/protocol/user");
const typeRegistry_1 = require("../../../typeRegistry");
exports.protobufPackage = "xray.proxy.vmess.inbound";
function createBaseDetourConfig() {
    return { $type: "xray.proxy.vmess.inbound.DetourConfig", to: "" };
}
exports.DetourConfig = {
    $type: "xray.proxy.vmess.inbound.DetourConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.to !== "") {
            writer.uint32(10).string(message.to);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDetourConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.to = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.DetourConfig.$type, to: isSet(object.to) ? globalThis.String(object.to) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.to !== "") {
            obj.to = message.to;
        }
        return obj;
    },
    create(base) {
        return exports.DetourConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseDetourConfig();
        message.to = object.to ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.DetourConfig.$type, exports.DetourConfig);
function createBaseDefaultConfig() {
    return { $type: "xray.proxy.vmess.inbound.DefaultConfig", level: 0 };
}
exports.DefaultConfig = {
    $type: "xray.proxy.vmess.inbound.DefaultConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.level !== 0) {
            writer.uint32(16).uint32(message.level);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDefaultConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.level = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.DefaultConfig.$type, level: isSet(object.level) ? globalThis.Number(object.level) : 0 };
    },
    toJSON(message) {
        const obj = {};
        if (message.level !== 0) {
            obj.level = Math.round(message.level);
        }
        return obj;
    },
    create(base) {
        return exports.DefaultConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseDefaultConfig();
        message.level = object.level ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.DefaultConfig.$type, exports.DefaultConfig);
function createBaseConfig() {
    return { $type: "xray.proxy.vmess.inbound.Config", user: [], default: undefined };
}
exports.Config = {
    $type: "xray.proxy.vmess.inbound.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.user) {
            user_1.User.encode(v, writer.uint32(10).fork()).join();
        }
        if (message.default !== undefined) {
            exports.DefaultConfig.encode(message.default, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.user.push(user_1.User.decode(reader, reader.uint32()));
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.default = exports.DefaultConfig.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            user: globalThis.Array.isArray(object?.user) ? object.user.map((e) => user_1.User.fromJSON(e)) : [],
            default: isSet(object.default) ? exports.DefaultConfig.fromJSON(object.default) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.user?.length) {
            obj.user = message.user.map((e) => user_1.User.toJSON(e));
        }
        if (message.default !== undefined) {
            obj.default = exports.DefaultConfig.toJSON(message.default);
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.user = object.user?.map((e) => user_1.User.fromPartial(e)) || [];
        message.default = (object.default !== undefined && object.default !== null)
            ? exports.DefaultConfig.fromPartial(object.default)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function isSet(value) {
    return value !== null && value !== undefined;
}
