// @remnawave/xtls-sdk/build/src/xray-protos/proxy/shadowsocks_2022/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/shadowsocks_2022/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientConfig = exports.Account = exports.RelayServerConfig = exports.RelayDestination = exports.MultiUserServerConfig = exports.ServerConfig = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const address_1 = require("../../common/net/address");
const network_1 = require("../../common/net/network");
const user_1 = require("../../common/protocol/user");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.proxy.shadowsocks_2022";
function createBaseServerConfig() {
    return { $type: "xray.proxy.shadowsocks_2022.ServerConfig", method: "", key: "", email: "", level: 0, network: [] };
}
exports.ServerConfig = {
    $type: "xray.proxy.shadowsocks_2022.ServerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.method !== "") {
            writer.uint32(10).string(message.method);
        }
        if (message.key !== "") {
            writer.uint32(18).string(message.key);
        }
        if (message.email !== "") {
            writer.uint32(26).string(message.email);
        }
        if (message.level !== 0) {
            writer.uint32(32).int32(message.level);
        }
        writer.uint32(42).fork();
        for (const v of message.network) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseServerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.method = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.email = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.level = reader.int32();
                    continue;
                }
                case 5: {
                    if (tag === 40) {
                        message.network.push(reader.int32());
                        continue;
                    }
                    if (tag === 42) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.network.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ServerConfig.$type,
            method: isSet(object.method) ? globalThis.String(object.method) : "",
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            email: isSet(object.email) ? globalThis.String(object.email) : "",
            level: isSet(object.level) ? globalThis.Number(object.level) : 0,
            network: globalThis.Array.isArray(object?.network) ? object.network.map((e) => (0, network_1.networkFromJSON)(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.method !== "") {
            obj.method = message.method;
        }
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.email !== "") {
            obj.email = message.email;
        }
        if (message.level !== 0) {
            obj.level = Math.round(message.level);
        }
        if (message.network?.length) {
            obj.network = message.network.map((e) => (0, network_1.networkToJSON)(e));
        }
        return obj;
    },
    create(base) {
        return exports.ServerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseServerConfig();
        message.method = object.method ?? "";
        message.key = object.key ?? "";
        message.email = object.email ?? "";
        message.level = object.level ?? 0;
        message.network = object.network?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ServerConfig.$type, exports.ServerConfig);
function createBaseMultiUserServerConfig() {
    return { $type: "xray.proxy.shadowsocks_2022.MultiUserServerConfig", method: "", key: "", users: [], network: [] };
}
exports.MultiUserServerConfig = {
    $type: "xray.proxy.shadowsocks_2022.MultiUserServerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.method !== "") {
            writer.uint32(10).string(message.method);
        }
        if (message.key !== "") {
            writer.uint32(18).string(message.key);
        }
        for (const v of message.users) {
            user_1.User.encode(v, writer.uint32(26).fork()).join();
        }
        writer.uint32(34).fork();
        for (const v of message.network) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMultiUserServerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.method = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.users.push(user_1.User.decode(reader, reader.uint32()));
                    continue;
                }
                case 4: {
                    if (tag === 32) {
                        message.network.push(reader.int32());
                        continue;
                    }
                    if (tag === 34) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.network.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.MultiUserServerConfig.$type,
            method: isSet(object.method) ? globalThis.String(object.method) : "",
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            users: globalThis.Array.isArray(object?.users) ? object.users.map((e) => user_1.User.fromJSON(e)) : [],
            network: globalThis.Array.isArray(object?.network) ? object.network.map((e) => (0, network_1.networkFromJSON)(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.method !== "") {
            obj.method = message.method;
        }
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.users?.length) {
            obj.users = message.users.map((e) => user_1.User.toJSON(e));
        }
        if (message.network?.length) {
            obj.network = message.network.map((e) => (0, network_1.networkToJSON)(e));
        }
        return obj;
    },
    create(base) {
        return exports.MultiUserServerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseMultiUserServerConfig();
        message.method = object.method ?? "";
        message.key = object.key ?? "";
        message.users = object.users?.map((e) => user_1.User.fromPartial(e)) || [];
        message.network = object.network?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.MultiUserServerConfig.$type, exports.MultiUserServerConfig);
function createBaseRelayDestination() {
    return {
        $type: "xray.proxy.shadowsocks_2022.RelayDestination",
        key: "",
        address: undefined,
        port: 0,
        email: "",
        level: 0,
    };
}
exports.RelayDestination = {
    $type: "xray.proxy.shadowsocks_2022.RelayDestination",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.address !== undefined) {
            address_1.IPOrDomain.encode(message.address, writer.uint32(18).fork()).join();
        }
        if (message.port !== 0) {
            writer.uint32(24).uint32(message.port);
        }
        if (message.email !== "") {
            writer.uint32(34).string(message.email);
        }
        if (message.level !== 0) {
            writer.uint32(40).int32(message.level);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRelayDestination();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.address = address_1.IPOrDomain.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.port = reader.uint32();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.email = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.level = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.RelayDestination.$type,
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            address: isSet(object.address) ? address_1.IPOrDomain.fromJSON(object.address) : undefined,
            port: isSet(object.port) ? globalThis.Number(object.port) : 0,
            email: isSet(object.email) ? globalThis.String(object.email) : "",
            level: isSet(object.level) ? globalThis.Number(object.level) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.address !== undefined) {
            obj.address = address_1.IPOrDomain.toJSON(message.address);
        }
        if (message.port !== 0) {
            obj.port = Math.round(message.port);
        }
        if (message.email !== "") {
            obj.email = message.email;
        }
        if (message.level !== 0) {
            obj.level = Math.round(message.level);
        }
        return obj;
    },
    create(base) {
        return exports.RelayDestination.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRelayDestination();
        message.key = object.key ?? "";
        message.address = (object.address !== undefined && object.address !== null)
            ? address_1.IPOrDomain.fromPartial(object.address)
            : undefined;
        message.port = object.port ?? 0;
        message.email = object.email ?? "";
        message.level = object.level ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.RelayDestination.$type, exports.RelayDestination);
function createBaseRelayServerConfig() {
    return { $type: "xray.proxy.shadowsocks_2022.RelayServerConfig", method: "", key: "", destinations: [], network: [] };
}
exports.RelayServerConfig = {
    $type: "xray.proxy.shadowsocks_2022.RelayServerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.method !== "") {
            writer.uint32(10).string(message.method);
        }
        if (message.key !== "") {
            writer.uint32(18).string(message.key);
        }
        for (const v of message.destinations) {
            exports.RelayDestination.encode(v, writer.uint32(26).fork()).join();
        }
        writer.uint32(34).fork();
        for (const v of message.network) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRelayServerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.method = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.destinations.push(exports.RelayDestination.decode(reader, reader.uint32()));
                    continue;
                }
                case 4: {
                    if (tag === 32) {
                        message.network.push(reader.int32());
                        continue;
                    }
                    if (tag === 34) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.network.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.RelayServerConfig.$type,
            method: isSet(object.method) ? globalThis.String(object.method) : "",
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            destinations: globalThis.Array.isArray(object?.destinations)
                ? object.destinations.map((e) => exports.RelayDestination.fromJSON(e))
                : [],
            network: globalThis.Array.isArray(object?.network) ? object.network.map((e) => (0, network_1.networkFromJSON)(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.method !== "") {
            obj.method = message.method;
        }
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.destinations?.length) {
            obj.destinations = message.destinations.map((e) => exports.RelayDestination.toJSON(e));
        }
        if (message.network?.length) {
            obj.network = message.network.map((e) => (0, network_1.networkToJSON)(e));
        }
        return obj;
    },
    create(base) {
        return exports.RelayServerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRelayServerConfig();
        message.method = object.method ?? "";
        message.key = object.key ?? "";
        message.destinations = object.destinations?.map((e) => exports.RelayDestination.fromPartial(e)) || [];
        message.network = object.network?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.RelayServerConfig.$type, exports.RelayServerConfig);
function createBaseAccount() {
    return { $type: "xray.proxy.shadowsocks_2022.Account", key: "" };
}
exports.Account = {
    $type: "xray.proxy.shadowsocks_2022.Account",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.Account.$type, key: isSet(object.key) ? globalThis.String(object.key) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        return obj;
    },
    create(base) {
        return exports.Account.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccount();
        message.key = object.key ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Account.$type, exports.Account);
function createBaseClientConfig() {
    return {
        $type: "xray.proxy.shadowsocks_2022.ClientConfig",
        address: undefined,
        port: 0,
        method: "",
        key: "",
        udpOverTcp: false,
        udpOverTcpVersion: 0,
    };
}
exports.ClientConfig = {
    $type: "xray.proxy.shadowsocks_2022.ClientConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.address !== undefined) {
            address_1.IPOrDomain.encode(message.address, writer.uint32(10).fork()).join();
        }
        if (message.port !== 0) {
            writer.uint32(16).uint32(message.port);
        }
        if (message.method !== "") {
            writer.uint32(26).string(message.method);
        }
        if (message.key !== "") {
            writer.uint32(34).string(message.key);
        }
        if (message.udpOverTcp !== false) {
            writer.uint32(40).bool(message.udpOverTcp);
        }
        if (message.udpOverTcpVersion !== 0) {
            writer.uint32(48).uint32(message.udpOverTcpVersion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseClientConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.address = address_1.IPOrDomain.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.port = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.method = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.udpOverTcp = reader.bool();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.udpOverTcpVersion = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ClientConfig.$type,
            address: isSet(object.address) ? address_1.IPOrDomain.fromJSON(object.address) : undefined,
            port: isSet(object.port) ? globalThis.Number(object.port) : 0,
            method: isSet(object.method) ? globalThis.String(object.method) : "",
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            udpOverTcp: isSet(object.udpOverTcp) ? globalThis.Boolean(object.udpOverTcp) : false,
            udpOverTcpVersion: isSet(object.udpOverTcpVersion) ? globalThis.Number(object.udpOverTcpVersion) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.address !== undefined) {
            obj.address = address_1.IPOrDomain.toJSON(message.address);
        }
        if (message.port !== 0) {
            obj.port = Math.round(message.port);
        }
        if (message.method !== "") {
            obj.method = message.method;
        }
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.udpOverTcp !== false) {
            obj.udpOverTcp = message.udpOverTcp;
        }
        if (message.udpOverTcpVersion !== 0) {
            obj.udpOverTcpVersion = Math.round(message.udpOverTcpVersion);
        }
        return obj;
    },
    create(base) {
        return exports.ClientConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseClientConfig();
        message.address = (object.address !== undefined && object.address !== null)
            ? address_1.IPOrDomain.fromPartial(object.address)
            : undefined;
        message.port = object.port ?? 0;
        message.method = object.method ?? "";
        message.key = object.key ?? "";
        message.udpOverTcp = object.udpOverTcp ?? false;
        message.udpOverTcpVersion = object.udpOverTcpVersion ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ClientConfig.$type, exports.ClientConfig);
function isSet(value) {
    return value !== null && value !== undefined;
}
