// @remnawave/xtls-sdk/build/src/xray-protos/proxy/dns/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/dns/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const destination_1 = require("../../common/net/destination");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.proxy.dns";
function createBaseConfig() {
    return { $type: "xray.proxy.dns.Config", server: undefined, userLevel: 0, nonIPQuery: "", blockTypes: [] };
}
exports.Config = {
    $type: "xray.proxy.dns.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.server !== undefined) {
            destination_1.Endpoint.encode(message.server, writer.uint32(10).fork()).join();
        }
        if (message.userLevel !== 0) {
            writer.uint32(16).uint32(message.userLevel);
        }
        if (message.nonIPQuery !== "") {
            writer.uint32(26).string(message.nonIPQuery);
        }
        writer.uint32(34).fork();
        for (const v of message.blockTypes) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.server = destination_1.Endpoint.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.userLevel = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.nonIPQuery = reader.string();
                    continue;
                }
                case 4: {
                    if (tag === 32) {
                        message.blockTypes.push(reader.int32());
                        continue;
                    }
                    if (tag === 34) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.blockTypes.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            server: isSet(object.server) ? destination_1.Endpoint.fromJSON(object.server) : undefined,
            userLevel: isSet(object.userLevel) ? globalThis.Number(object.userLevel) : 0,
            nonIPQuery: isSet(object.nonIPQuery) ? globalThis.String(object.nonIPQuery) : "",
            blockTypes: globalThis.Array.isArray(object?.blockTypes)
                ? object.blockTypes.map((e) => globalThis.Number(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.server !== undefined) {
            obj.server = destination_1.Endpoint.toJSON(message.server);
        }
        if (message.userLevel !== 0) {
            obj.userLevel = Math.round(message.userLevel);
        }
        if (message.nonIPQuery !== "") {
            obj.nonIPQuery = message.nonIPQuery;
        }
        if (message.blockTypes?.length) {
            obj.blockTypes = message.blockTypes.map((e) => Math.round(e));
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.server = (object.server !== undefined && object.server !== null)
            ? destination_1.Endpoint.fromPartial(object.server)
            : undefined;
        message.userLevel = object.userLevel ?? 0;
        message.nonIPQuery = object.nonIPQuery ?? "";
        message.blockTypes = object.blockTypes?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function isSet(value) {
    return value !== null && value !== undefined;
}
