// @remnawave/xtls-sdk/build/src/xray-protos/proxy/shadowsocks/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/shadowsocks/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientConfig = exports.ServerConfig = exports.Account = exports.CipherType = exports.protobufPackage = void 0;
exports.cipherTypeFromJSON = cipherTypeFromJSON;
exports.cipherTypeToJSON = cipherTypeToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const network_1 = require("../../common/net/network");
const server_spec_1 = require("../../common/protocol/server_spec");
const user_1 = require("../../common/protocol/user");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.proxy.shadowsocks";
var CipherType;
(function (CipherType) {
    CipherType[CipherType["UNKNOWN"] = 0] = "UNKNOWN";
    CipherType[CipherType["AES_128_GCM"] = 5] = "AES_128_GCM";
    CipherType[CipherType["AES_256_GCM"] = 6] = "AES_256_GCM";
    CipherType[CipherType["CHACHA20_POLY1305"] = 7] = "CHACHA20_POLY1305";
    CipherType[CipherType["XCHACHA20_POLY1305"] = 8] = "XCHACHA20_POLY1305";
    CipherType[CipherType["NONE"] = 9] = "NONE";
    CipherType[CipherType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(CipherType || (exports.CipherType = CipherType = {}));
function cipherTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return CipherType.UNKNOWN;
        case 5:
        case "AES_128_GCM":
            return CipherType.AES_128_GCM;
        case 6:
        case "AES_256_GCM":
            return CipherType.AES_256_GCM;
        case 7:
        case "CHACHA20_POLY1305":
            return CipherType.CHACHA20_POLY1305;
        case 8:
        case "XCHACHA20_POLY1305":
            return CipherType.XCHACHA20_POLY1305;
        case 9:
        case "NONE":
            return CipherType.NONE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return CipherType.UNRECOGNIZED;
    }
}
function cipherTypeToJSON(object) {
    switch (object) {
        case CipherType.UNKNOWN:
            return "UNKNOWN";
        case CipherType.AES_128_GCM:
            return "AES_128_GCM";
        case CipherType.AES_256_GCM:
            return "AES_256_GCM";
        case CipherType.CHACHA20_POLY1305:
            return "CHACHA20_POLY1305";
        case CipherType.XCHACHA20_POLY1305:
            return "XCHACHA20_POLY1305";
        case CipherType.NONE:
            return "NONE";
        case CipherType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseAccount() {
    return { $type: "xray.proxy.shadowsocks.Account", password: "", cipherType: 0, ivCheck: false };
}
exports.Account = {
    $type: "xray.proxy.shadowsocks.Account",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.password !== "") {
            writer.uint32(10).string(message.password);
        }
        if (message.cipherType !== 0) {
            writer.uint32(16).int32(message.cipherType);
        }
        if (message.ivCheck !== false) {
            writer.uint32(24).bool(message.ivCheck);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.password = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.cipherType = reader.int32();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.ivCheck = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Account.$type,
            password: isSet(object.password) ? globalThis.String(object.password) : "",
            cipherType: isSet(object.cipherType) ? cipherTypeFromJSON(object.cipherType) : 0,
            ivCheck: isSet(object.ivCheck) ? globalThis.Boolean(object.ivCheck) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.password !== "") {
            obj.password = message.password;
        }
        if (message.cipherType !== 0) {
            obj.cipherType = cipherTypeToJSON(message.cipherType);
        }
        if (message.ivCheck !== false) {
            obj.ivCheck = message.ivCheck;
        }
        return obj;
    },
    create(base) {
        return exports.Account.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccount();
        message.password = object.password ?? "";
        message.cipherType = object.cipherType ?? 0;
        message.ivCheck = object.ivCheck ?? false;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Account.$type, exports.Account);
function createBaseServerConfig() {
    return { $type: "xray.proxy.shadowsocks.ServerConfig", users: [], network: [] };
}
exports.ServerConfig = {
    $type: "xray.proxy.shadowsocks.ServerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.users) {
            user_1.User.encode(v, writer.uint32(10).fork()).join();
        }
        writer.uint32(18).fork();
        for (const v of message.network) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseServerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.users.push(user_1.User.decode(reader, reader.uint32()));
                    continue;
                }
                case 2: {
                    if (tag === 16) {
                        message.network.push(reader.int32());
                        continue;
                    }
                    if (tag === 18) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.network.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ServerConfig.$type,
            users: globalThis.Array.isArray(object?.users) ? object.users.map((e) => user_1.User.fromJSON(e)) : [],
            network: globalThis.Array.isArray(object?.network) ? object.network.map((e) => (0, network_1.networkFromJSON)(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.users?.length) {
            obj.users = message.users.map((e) => user_1.User.toJSON(e));
        }
        if (message.network?.length) {
            obj.network = message.network.map((e) => (0, network_1.networkToJSON)(e));
        }
        return obj;
    },
    create(base) {
        return exports.ServerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseServerConfig();
        message.users = object.users?.map((e) => user_1.User.fromPartial(e)) || [];
        message.network = object.network?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ServerConfig.$type, exports.ServerConfig);
function createBaseClientConfig() {
    return { $type: "xray.proxy.shadowsocks.ClientConfig", server: undefined };
}
exports.ClientConfig = {
    $type: "xray.proxy.shadowsocks.ClientConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.server !== undefined) {
            server_spec_1.ServerEndpoint.encode(message.server, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseClientConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.server = server_spec_1.ServerEndpoint.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ClientConfig.$type,
            server: isSet(object.server) ? server_spec_1.ServerEndpoint.fromJSON(object.server) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.server !== undefined) {
            obj.server = server_spec_1.ServerEndpoint.toJSON(message.server);
        }
        return obj;
    },
    create(base) {
        return exports.ClientConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseClientConfig();
        message.server = (object.server !== undefined && object.server !== null)
            ? server_spec_1.ServerEndpoint.fromPartial(object.server)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ClientConfig.$type, exports.ClientConfig);
function isSet(value) {
    return value !== null && value !== undefined;
}
