// @remnawave/xtls-sdk/build/src/xray-protos/common/log/log.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/log/log.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Severity = exports.protobufPackage = void 0;
exports.severityFromJSON = severityFromJSON;
exports.severityToJSON = severityToJSON;
/* eslint-disable */
exports.protobufPackage = "xray.common.log";
var Severity;
(function (Severity) {
    Severity[Severity["Unknown"] = 0] = "Unknown";
    Severity[Severity["Error"] = 1] = "Error";
    Severity[Severity["Warning"] = 2] = "Warning";
    Severity[Severity["Info"] = 3] = "Info";
    Severity[Severity["Debug"] = 4] = "Debug";
    Severity[Severity["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Severity || (exports.Severity = Severity = {}));
function severityFromJSON(object) {
    switch (object) {
        case 0:
        case "Unknown":
            return Severity.Unknown;
        case 1:
        case "Error":
            return Severity.Error;
        case 2:
        case "Warning":
            return Severity.Warning;
        case 3:
        case "Info":
            return Severity.Info;
        case 4:
        case "Debug":
            return Severity.Debug;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Severity.UNRECOGNIZED;
    }
}
function severityToJSON(object) {
    switch (object) {
        case Severity.Unknown:
            return "Unknown";
        case Severity.Error:
            return "Error";
        case Severity.Warning:
            return "Warning";
        case Severity.Info:
            return "Info";
        case Severity.Debug:
            return "Debug";
        case Severity.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
