// @remnawave/xtls-sdk/build/src/xray-protos/common/protocol/headers.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/protocol/headers.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityConfig = exports.SecurityType = exports.protobufPackage = void 0;
exports.securityTypeFromJSON = securityTypeFromJSON;
exports.securityTypeToJSON = securityTypeToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.common.protocol";
var SecurityType;
(function (SecurityType) {
    SecurityType[SecurityType["UNKNOWN"] = 0] = "UNKNOWN";
    SecurityType[SecurityType["AUTO"] = 2] = "AUTO";
    SecurityType[SecurityType["AES128_GCM"] = 3] = "AES128_GCM";
    SecurityType[SecurityType["CHACHA20_POLY1305"] = 4] = "CHACHA20_POLY1305";
    /** NONE - [DEPRECATED 2023-06] */
    SecurityType[SecurityType["NONE"] = 5] = "NONE";
    SecurityType[SecurityType["ZERO"] = 6] = "ZERO";
    SecurityType[SecurityType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SecurityType || (exports.SecurityType = SecurityType = {}));
function securityTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return SecurityType.UNKNOWN;
        case 2:
        case "AUTO":
            return SecurityType.AUTO;
        case 3:
        case "AES128_GCM":
            return SecurityType.AES128_GCM;
        case 4:
        case "CHACHA20_POLY1305":
            return SecurityType.CHACHA20_POLY1305;
        case 5:
        case "NONE":
            return SecurityType.NONE;
        case 6:
        case "ZERO":
            return SecurityType.ZERO;
        case -1:
        case "UNRECOGNIZED":
        default:
            return SecurityType.UNRECOGNIZED;
    }
}
function securityTypeToJSON(object) {
    switch (object) {
        case SecurityType.UNKNOWN:
            return "UNKNOWN";
        case SecurityType.AUTO:
            return "AUTO";
        case SecurityType.AES128_GCM:
            return "AES128_GCM";
        case SecurityType.CHACHA20_POLY1305:
            return "CHACHA20_POLY1305";
        case SecurityType.NONE:
            return "NONE";
        case SecurityType.ZERO:
            return "ZERO";
        case SecurityType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseSecurityConfig() {
    return { $type: "xray.common.protocol.SecurityConfig", type: 0 };
}
exports.SecurityConfig = {
    $type: "xray.common.protocol.SecurityConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.type !== 0) {
            writer.uint32(8).int32(message.type);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSecurityConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.type = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.SecurityConfig.$type, type: isSet(object.type) ? securityTypeFromJSON(object.type) : 0 };
    },
    toJSON(message) {
        const obj = {};
        if (message.type !== 0) {
            obj.type = securityTypeToJSON(message.type);
        }
        return obj;
    },
    create(base) {
        return exports.SecurityConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSecurityConfig();
        message.type = object.type ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.SecurityConfig.$type, exports.SecurityConfig);
function isSet(value) {
    return value !== null && value !== undefined;
}
