// @remnawave/xtls-sdk/build/src/xray-protos/common/protocol/server_spec.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/protocol/server_spec.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerEndpoint = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
const address_1 = require("../net/address");
const user_1 = require("./user");
exports.protobufPackage = "xray.common.protocol";
function createBaseServerEndpoint() {
    return { $type: "xray.common.protocol.ServerEndpoint", address: undefined, port: 0, user: undefined };
}
exports.ServerEndpoint = {
    $type: "xray.common.protocol.ServerEndpoint",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.address !== undefined) {
            address_1.IPOrDomain.encode(message.address, writer.uint32(10).fork()).join();
        }
        if (message.port !== 0) {
            writer.uint32(16).uint32(message.port);
        }
        if (message.user !== undefined) {
            user_1.User.encode(message.user, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseServerEndpoint();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.address = address_1.IPOrDomain.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.port = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.user = user_1.User.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ServerEndpoint.$type,
            address: isSet(object.address) ? address_1.IPOrDomain.fromJSON(object.address) : undefined,
            port: isSet(object.port) ? globalThis.Number(object.port) : 0,
            user: isSet(object.user) ? user_1.User.fromJSON(object.user) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.address !== undefined) {
            obj.address = address_1.IPOrDomain.toJSON(message.address);
        }
        if (message.port !== 0) {
            obj.port = Math.round(message.port);
        }
        if (message.user !== undefined) {
            obj.user = user_1.User.toJSON(message.user);
        }
        return obj;
    },
    create(base) {
        return exports.ServerEndpoint.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseServerEndpoint();
        message.address = (object.address !== undefined && object.address !== null)
            ? address_1.IPOrDomain.fromPartial(object.address)
            : undefined;
        message.port = object.port ?? 0;
        message.user = (object.user !== undefined && object.user !== null) ? user_1.User.fromPartial(object.user) : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ServerEndpoint.$type, exports.ServerEndpoint);
function isSet(value) {
    return value !== null && value !== undefined;
}
