// @remnawave/xtls-sdk/build/src/xray-protos/common/protocol/user.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: common/protocol/user.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
const typed_message_1 = require("../serial/typed_message");
exports.protobufPackage = "xray.common.protocol";
function createBaseUser() {
    return { $type: "xray.common.protocol.User", level: 0, email: "", account: undefined };
}
exports.User = {
    $type: "xray.common.protocol.User",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.level !== 0) {
            writer.uint32(8).uint32(message.level);
        }
        if (message.email !== "") {
            writer.uint32(18).string(message.email);
        }
        if (message.account !== undefined) {
            typed_message_1.TypedMessage.encode(message.account, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUser();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.level = reader.uint32();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.email = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.account = typed_message_1.TypedMessage.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.User.$type,
            level: isSet(object.level) ? globalThis.Number(object.level) : 0,
            email: isSet(object.email) ? globalThis.String(object.email) : "",
            account: isSet(object.account) ? typed_message_1.TypedMessage.fromJSON(object.account) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.level !== 0) {
            obj.level = Math.round(message.level);
        }
        if (message.email !== "") {
            obj.email = message.email;
        }
        if (message.account !== undefined) {
            obj.account = typed_message_1.TypedMessage.toJSON(message.account);
        }
        return obj;
    },
    create(base) {
        return exports.User.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseUser();
        message.level = object.level ?? 0;
        message.email = object.email ?? "";
        message.account = (object.account !== undefined && object.account !== null)
            ? typed_message_1.TypedMessage.fromPartial(object.account)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.User.$type, exports.User);
function isSet(value) {
    return value !== null && value !== undefined;
}
