// @remnawave/xtls-sdk/build/src/xray-protos/app/commander/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/commander/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReflectionConfig = exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typed_message_1 = require("../../common/serial/typed_message");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.commander";
function createBaseConfig() {
    return { $type: "xray.app.commander.Config", tag: "", listen: "", service: [] };
}
exports.Config = {
    $type: "xray.app.commander.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tag !== "") {
            writer.uint32(10).string(message.tag);
        }
        if (message.listen !== "") {
            writer.uint32(26).string(message.listen);
        }
        for (const v of message.service) {
            typed_message_1.TypedMessage.encode(v, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.listen = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.service.push(typed_message_1.TypedMessage.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            tag: isSet(object.tag) ? globalThis.String(object.tag) : "",
            listen: isSet(object.listen) ? globalThis.String(object.listen) : "",
            service: globalThis.Array.isArray(object?.service)
                ? object.service.map((e) => typed_message_1.TypedMessage.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.tag !== "") {
            obj.tag = message.tag;
        }
        if (message.listen !== "") {
            obj.listen = message.listen;
        }
        if (message.service?.length) {
            obj.service = message.service.map((e) => typed_message_1.TypedMessage.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.tag = object.tag ?? "";
        message.listen = object.listen ?? "";
        message.service = object.service?.map((e) => typed_message_1.TypedMessage.fromPartial(e)) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function createBaseReflectionConfig() {
    return { $type: "xray.app.commander.ReflectionConfig" };
}
exports.ReflectionConfig = {
    $type: "xray.app.commander.ReflectionConfig",
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReflectionConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return { $type: exports.ReflectionConfig.$type };
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.ReflectionConfig.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseReflectionConfig();
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ReflectionConfig.$type, exports.ReflectionConfig);
function isSet(value) {
    return value !== null && value !== undefined;
}
