// @remnawave/xtls-sdk/build/src/xray-protos/proxy/vless/outbound/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/vless/outbound/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const server_spec_1 = require("../../../common/protocol/server_spec");
const typeRegistry_1 = require("../../../typeRegistry");
exports.protobufPackage = "xray.proxy.vless.outbound";
function createBaseConfig() {
    return { $type: "xray.proxy.vless.outbound.Config", vnext: undefined };
}
exports.Config = {
    $type: "xray.proxy.vless.outbound.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.vnext !== undefined) {
            server_spec_1.ServerEndpoint.encode(message.vnext, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.vnext = server_spec_1.ServerEndpoint.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.Config.$type, vnext: isSet(object.vnext) ? server_spec_1.ServerEndpoint.fromJSON(object.vnext) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.vnext !== undefined) {
            obj.vnext = server_spec_1.ServerEndpoint.toJSON(message.vnext);
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.vnext = (object.vnext !== undefined && object.vnext !== null)
            ? server_spec_1.ServerEndpoint.fromPartial(object.vnext)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function isSet(value) {
    return value !== null && value !== undefined;
}
