// @remnawave/xtls-sdk/build/src/xray-protos/app/log/command/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/log/command/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerServiceDefinition = exports.RestartLoggerResponse = exports.RestartLoggerRequest = exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../../typeRegistry");
exports.protobufPackage = "xray.app.log.command";
function createBaseConfig() {
    return { $type: "xray.app.log.command.Config" };
}
exports.Config = {
    $type: "xray.app.log.command.Config",
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return { $type: exports.Config.$type };
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseConfig();
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function createBaseRestartLoggerRequest() {
    return { $type: "xray.app.log.command.RestartLoggerRequest" };
}
exports.RestartLoggerRequest = {
    $type: "xray.app.log.command.RestartLoggerRequest",
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRestartLoggerRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return { $type: exports.RestartLoggerRequest.$type };
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RestartLoggerRequest.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseRestartLoggerRequest();
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.RestartLoggerRequest.$type, exports.RestartLoggerRequest);
function createBaseRestartLoggerResponse() {
    return { $type: "xray.app.log.command.RestartLoggerResponse" };
}
exports.RestartLoggerResponse = {
    $type: "xray.app.log.command.RestartLoggerResponse",
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRestartLoggerResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return { $type: exports.RestartLoggerResponse.$type };
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RestartLoggerResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseRestartLoggerResponse();
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.RestartLoggerResponse.$type, exports.RestartLoggerResponse);
exports.LoggerServiceDefinition = {
    name: "LoggerService",
    fullName: "xray.app.log.command.LoggerService",
    methods: {
        restartLogger: {
            name: "RestartLogger",
            requestType: exports.RestartLoggerRequest,
            requestStream: false,
            responseType: exports.RestartLoggerResponse,
            responseStream: false,
            options: {},
        },
    },
};
