// @remnawave/xtls-sdk/build/src/xray-protos/app/dispatcher/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/dispatcher/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.SessionConfig = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.dispatcher";
function createBaseSessionConfig() {
    return { $type: "xray.app.dispatcher.SessionConfig" };
}
exports.SessionConfig = {
    $type: "xray.app.dispatcher.SessionConfig",
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSessionConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return { $type: exports.SessionConfig.$type };
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.SessionConfig.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseSessionConfig();
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.SessionConfig.$type, exports.SessionConfig);
function createBaseConfig() {
    return { $type: "xray.app.dispatcher.Config", settings: undefined };
}
exports.Config = {
    $type: "xray.app.dispatcher.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.settings !== undefined) {
            exports.SessionConfig.encode(message.settings, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.settings = exports.SessionConfig.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            settings: isSet(object.settings) ? exports.SessionConfig.fromJSON(object.settings) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.settings !== undefined) {
            obj.settings = exports.SessionConfig.toJSON(message.settings);
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.settings = (object.settings !== undefined && object.settings !== null)
            ? exports.SessionConfig.fromPartial(object.settings)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function isSet(value) {
    return value !== null && value !== undefined;
}
