// @remnawave/xtls-sdk/build/src/xray-protos/proxy/trojan/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/trojan/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerConfig = exports.ClientConfig = exports.Fallback = exports.Account = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const server_spec_1 = require("../../common/protocol/server_spec");
const user_1 = require("../../common/protocol/user");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.proxy.trojan";
function createBaseAccount() {
    return { $type: "xray.proxy.trojan.Account", password: "" };
}
exports.Account = {
    $type: "xray.proxy.trojan.Account",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.password !== "") {
            writer.uint32(10).string(message.password);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.password = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.Account.$type, password: isSet(object.password) ? globalThis.String(object.password) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.password !== "") {
            obj.password = message.password;
        }
        return obj;
    },
    create(base) {
        return exports.Account.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccount();
        message.password = object.password ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Account.$type, exports.Account);
function createBaseFallback() {
    return { $type: "xray.proxy.trojan.Fallback", name: "", alpn: "", path: "", type: "", dest: "", xver: 0 };
}
exports.Fallback = {
    $type: "xray.proxy.trojan.Fallback",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.alpn !== "") {
            writer.uint32(18).string(message.alpn);
        }
        if (message.path !== "") {
            writer.uint32(26).string(message.path);
        }
        if (message.type !== "") {
            writer.uint32(34).string(message.type);
        }
        if (message.dest !== "") {
            writer.uint32(42).string(message.dest);
        }
        if (message.xver !== 0) {
            writer.uint32(48).uint64(message.xver);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFallback();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.alpn = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.path = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.type = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.dest = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.xver = longToNumber(reader.uint64());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Fallback.$type,
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            alpn: isSet(object.alpn) ? globalThis.String(object.alpn) : "",
            path: isSet(object.path) ? globalThis.String(object.path) : "",
            type: isSet(object.type) ? globalThis.String(object.type) : "",
            dest: isSet(object.dest) ? globalThis.String(object.dest) : "",
            xver: isSet(object.xver) ? globalThis.Number(object.xver) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.alpn !== "") {
            obj.alpn = message.alpn;
        }
        if (message.path !== "") {
            obj.path = message.path;
        }
        if (message.type !== "") {
            obj.type = message.type;
        }
        if (message.dest !== "") {
            obj.dest = message.dest;
        }
        if (message.xver !== 0) {
            obj.xver = Math.round(message.xver);
        }
        return obj;
    },
    create(base) {
        return exports.Fallback.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseFallback();
        message.name = object.name ?? "";
        message.alpn = object.alpn ?? "";
        message.path = object.path ?? "";
        message.type = object.type ?? "";
        message.dest = object.dest ?? "";
        message.xver = object.xver ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Fallback.$type, exports.Fallback);
function createBaseClientConfig() {
    return { $type: "xray.proxy.trojan.ClientConfig", server: undefined };
}
exports.ClientConfig = {
    $type: "xray.proxy.trojan.ClientConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.server !== undefined) {
            server_spec_1.ServerEndpoint.encode(message.server, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseClientConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.server = server_spec_1.ServerEndpoint.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ClientConfig.$type,
            server: isSet(object.server) ? server_spec_1.ServerEndpoint.fromJSON(object.server) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.server !== undefined) {
            obj.server = server_spec_1.ServerEndpoint.toJSON(message.server);
        }
        return obj;
    },
    create(base) {
        return exports.ClientConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseClientConfig();
        message.server = (object.server !== undefined && object.server !== null)
            ? server_spec_1.ServerEndpoint.fromPartial(object.server)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ClientConfig.$type, exports.ClientConfig);
function createBaseServerConfig() {
    return { $type: "xray.proxy.trojan.ServerConfig", users: [], fallbacks: [] };
}
exports.ServerConfig = {
    $type: "xray.proxy.trojan.ServerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.users) {
            user_1.User.encode(v, writer.uint32(10).fork()).join();
        }
        for (const v of message.fallbacks) {
            exports.Fallback.encode(v, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseServerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.users.push(user_1.User.decode(reader, reader.uint32()));
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.fallbacks.push(exports.Fallback.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ServerConfig.$type,
            users: globalThis.Array.isArray(object?.users) ? object.users.map((e) => user_1.User.fromJSON(e)) : [],
            fallbacks: globalThis.Array.isArray(object?.fallbacks)
                ? object.fallbacks.map((e) => exports.Fallback.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.users?.length) {
            obj.users = message.users.map((e) => user_1.User.toJSON(e));
        }
        if (message.fallbacks?.length) {
            obj.fallbacks = message.fallbacks.map((e) => exports.Fallback.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.ServerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseServerConfig();
        message.users = object.users?.map((e) => user_1.User.fromPartial(e)) || [];
        message.fallbacks = object.fallbacks?.map((e) => exports.Fallback.fromPartial(e)) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ServerConfig.$type, exports.ServerConfig);
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
