// @remnawave/xtls-sdk/build/src/xray-protos/proxy/vless/encoding/addons.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/vless/encoding/addons.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Addons = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../../typeRegistry");
exports.protobufPackage = "xray.proxy.vless.encoding";
function createBaseAddons() {
    return { $type: "xray.proxy.vless.encoding.Addons", Flow: "", Seed: new Uint8Array(0) };
}
exports.Addons = {
    $type: "xray.proxy.vless.encoding.Addons",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.Flow !== "") {
            writer.uint32(10).string(message.Flow);
        }
        if (message.Seed.length !== 0) {
            writer.uint32(18).bytes(message.Seed);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddons();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.Flow = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.Seed = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Addons.$type,
            Flow: isSet(object.Flow) ? globalThis.String(object.Flow) : "",
            Seed: isSet(object.Seed) ? bytesFromBase64(object.Seed) : new Uint8Array(0),
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.Flow !== "") {
            obj.Flow = message.Flow;
        }
        if (message.Seed.length !== 0) {
            obj.Seed = base64FromBytes(message.Seed);
        }
        return obj;
    },
    create(base) {
        return exports.Addons.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAddons();
        message.Flow = object.Flow ?? "";
        message.Seed = object.Seed ?? new Uint8Array(0);
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Addons.$type, exports.Addons);
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
