// @remnawave/xtls-sdk/build/src/xray-protos/proxy/http/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/http/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientConfig = exports.Header = exports.ServerConfig_AccountsEntry = exports.ServerConfig = exports.Account = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const server_spec_1 = require("../../common/protocol/server_spec");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.proxy.http";
function createBaseAccount() {
    return { $type: "xray.proxy.http.Account", username: "", password: "" };
}
exports.Account = {
    $type: "xray.proxy.http.Account",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.username !== "") {
            writer.uint32(10).string(message.username);
        }
        if (message.password !== "") {
            writer.uint32(18).string(message.password);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.username = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.password = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Account.$type,
            username: isSet(object.username) ? globalThis.String(object.username) : "",
            password: isSet(object.password) ? globalThis.String(object.password) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.username !== "") {
            obj.username = message.username;
        }
        if (message.password !== "") {
            obj.password = message.password;
        }
        return obj;
    },
    create(base) {
        return exports.Account.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccount();
        message.username = object.username ?? "";
        message.password = object.password ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Account.$type, exports.Account);
function createBaseServerConfig() {
    return { $type: "xray.proxy.http.ServerConfig", accounts: {}, allowTransparent: false, userLevel: 0 };
}
exports.ServerConfig = {
    $type: "xray.proxy.http.ServerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        Object.entries(message.accounts).forEach(([key, value]) => {
            exports.ServerConfig_AccountsEntry.encode({ $type: "xray.proxy.http.ServerConfig.AccountsEntry", key: key, value }, writer.uint32(18).fork()).join();
        });
        if (message.allowTransparent !== false) {
            writer.uint32(24).bool(message.allowTransparent);
        }
        if (message.userLevel !== 0) {
            writer.uint32(32).uint32(message.userLevel);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseServerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    const entry2 = exports.ServerConfig_AccountsEntry.decode(reader, reader.uint32());
                    if (entry2.value !== undefined) {
                        message.accounts[entry2.key] = entry2.value;
                    }
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.allowTransparent = reader.bool();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.userLevel = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ServerConfig.$type,
            accounts: isObject(object.accounts)
                ? Object.entries(object.accounts).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
            allowTransparent: isSet(object.allowTransparent) ? globalThis.Boolean(object.allowTransparent) : false,
            userLevel: isSet(object.userLevel) ? globalThis.Number(object.userLevel) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.accounts) {
            const entries = Object.entries(message.accounts);
            if (entries.length > 0) {
                obj.accounts = {};
                entries.forEach(([k, v]) => {
                    obj.accounts[k] = v;
                });
            }
        }
        if (message.allowTransparent !== false) {
            obj.allowTransparent = message.allowTransparent;
        }
        if (message.userLevel !== 0) {
            obj.userLevel = Math.round(message.userLevel);
        }
        return obj;
    },
    create(base) {
        return exports.ServerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseServerConfig();
        message.accounts = Object.entries(object.accounts ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[key] = globalThis.String(value);
            }
            return acc;
        }, {});
        message.allowTransparent = object.allowTransparent ?? false;
        message.userLevel = object.userLevel ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ServerConfig.$type, exports.ServerConfig);
function createBaseServerConfig_AccountsEntry() {
    return { $type: "xray.proxy.http.ServerConfig.AccountsEntry", key: "", value: "" };
}
exports.ServerConfig_AccountsEntry = {
    $type: "xray.proxy.http.ServerConfig.AccountsEntry",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseServerConfig_AccountsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ServerConfig_AccountsEntry.$type,
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
    create(base) {
        return exports.ServerConfig_AccountsEntry.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseServerConfig_AccountsEntry();
        message.key = object.key ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ServerConfig_AccountsEntry.$type, exports.ServerConfig_AccountsEntry);
function createBaseHeader() {
    return { $type: "xray.proxy.http.Header", key: "", value: "" };
}
exports.Header = {
    $type: "xray.proxy.http.Header",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHeader();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Header.$type,
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
    create(base) {
        return exports.Header.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseHeader();
        message.key = object.key ?? "";
        message.value = object.value ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Header.$type, exports.Header);
function createBaseClientConfig() {
    return { $type: "xray.proxy.http.ClientConfig", server: undefined, header: [] };
}
exports.ClientConfig = {
    $type: "xray.proxy.http.ClientConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.server !== undefined) {
            server_spec_1.ServerEndpoint.encode(message.server, writer.uint32(10).fork()).join();
        }
        for (const v of message.header) {
            exports.Header.encode(v, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseClientConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.server = server_spec_1.ServerEndpoint.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.header.push(exports.Header.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.ClientConfig.$type,
            server: isSet(object.server) ? server_spec_1.ServerEndpoint.fromJSON(object.server) : undefined,
            header: globalThis.Array.isArray(object?.header) ? object.header.map((e) => exports.Header.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.server !== undefined) {
            obj.server = server_spec_1.ServerEndpoint.toJSON(message.server);
        }
        if (message.header?.length) {
            obj.header = message.header.map((e) => exports.Header.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.ClientConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseClientConfig();
        message.server = (object.server !== undefined && object.server !== null)
            ? server_spec_1.ServerEndpoint.fromPartial(object.server)
            : undefined;
        message.header = object.header?.map((e) => exports.Header.fromPartial(e)) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.ClientConfig.$type, exports.ClientConfig);
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
