// @remnawave/xtls-sdk/build/src/xray-protos/app/reverse/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/reverse/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.PortalConfig = exports.BridgeConfig = exports.Control = exports.Control_State = exports.protobufPackage = void 0;
exports.control_StateFromJSON = control_StateFromJSON;
exports.control_StateToJSON = control_StateToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.reverse";
var Control_State;
(function (Control_State) {
    Control_State[Control_State["ACTIVE"] = 0] = "ACTIVE";
    Control_State[Control_State["DRAIN"] = 1] = "DRAIN";
    Control_State[Control_State["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Control_State || (exports.Control_State = Control_State = {}));
function control_StateFromJSON(object) {
    switch (object) {
        case 0:
        case "ACTIVE":
            return Control_State.ACTIVE;
        case 1:
        case "DRAIN":
            return Control_State.DRAIN;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Control_State.UNRECOGNIZED;
    }
}
function control_StateToJSON(object) {
    switch (object) {
        case Control_State.ACTIVE:
            return "ACTIVE";
        case Control_State.DRAIN:
            return "DRAIN";
        case Control_State.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseControl() {
    return { $type: "xray.app.reverse.Control", state: 0, random: new Uint8Array(0) };
}
exports.Control = {
    $type: "xray.app.reverse.Control",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.state !== 0) {
            writer.uint32(8).int32(message.state);
        }
        if (message.random.length !== 0) {
            writer.uint32(794).bytes(message.random);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseControl();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.state = reader.int32();
                    continue;
                }
                case 99: {
                    if (tag !== 794) {
                        break;
                    }
                    message.random = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Control.$type,
            state: isSet(object.state) ? control_StateFromJSON(object.state) : 0,
            random: isSet(object.random) ? bytesFromBase64(object.random) : new Uint8Array(0),
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.state !== 0) {
            obj.state = control_StateToJSON(message.state);
        }
        if (message.random.length !== 0) {
            obj.random = base64FromBytes(message.random);
        }
        return obj;
    },
    create(base) {
        return exports.Control.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseControl();
        message.state = object.state ?? 0;
        message.random = object.random ?? new Uint8Array(0);
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Control.$type, exports.Control);
function createBaseBridgeConfig() {
    return { $type: "xray.app.reverse.BridgeConfig", tag: "", domain: "" };
}
exports.BridgeConfig = {
    $type: "xray.app.reverse.BridgeConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tag !== "") {
            writer.uint32(10).string(message.tag);
        }
        if (message.domain !== "") {
            writer.uint32(18).string(message.domain);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBridgeConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.domain = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.BridgeConfig.$type,
            tag: isSet(object.tag) ? globalThis.String(object.tag) : "",
            domain: isSet(object.domain) ? globalThis.String(object.domain) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.tag !== "") {
            obj.tag = message.tag;
        }
        if (message.domain !== "") {
            obj.domain = message.domain;
        }
        return obj;
    },
    create(base) {
        return exports.BridgeConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseBridgeConfig();
        message.tag = object.tag ?? "";
        message.domain = object.domain ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.BridgeConfig.$type, exports.BridgeConfig);
function createBasePortalConfig() {
    return { $type: "xray.app.reverse.PortalConfig", tag: "", domain: "" };
}
exports.PortalConfig = {
    $type: "xray.app.reverse.PortalConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tag !== "") {
            writer.uint32(10).string(message.tag);
        }
        if (message.domain !== "") {
            writer.uint32(18).string(message.domain);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePortalConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.domain = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.PortalConfig.$type,
            tag: isSet(object.tag) ? globalThis.String(object.tag) : "",
            domain: isSet(object.domain) ? globalThis.String(object.domain) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.tag !== "") {
            obj.tag = message.tag;
        }
        if (message.domain !== "") {
            obj.domain = message.domain;
        }
        return obj;
    },
    create(base) {
        return exports.PortalConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePortalConfig();
        message.tag = object.tag ?? "";
        message.domain = object.domain ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.PortalConfig.$type, exports.PortalConfig);
function createBaseConfig() {
    return { $type: "xray.app.reverse.Config", bridgeConfig: [], portalConfig: [] };
}
exports.Config = {
    $type: "xray.app.reverse.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.bridgeConfig) {
            exports.BridgeConfig.encode(v, writer.uint32(10).fork()).join();
        }
        for (const v of message.portalConfig) {
            exports.PortalConfig.encode(v, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.bridgeConfig.push(exports.BridgeConfig.decode(reader, reader.uint32()));
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.portalConfig.push(exports.PortalConfig.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            bridgeConfig: globalThis.Array.isArray(object?.bridgeConfig)
                ? object.bridgeConfig.map((e) => exports.BridgeConfig.fromJSON(e))
                : [],
            portalConfig: globalThis.Array.isArray(object?.portalConfig)
                ? object.portalConfig.map((e) => exports.PortalConfig.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bridgeConfig?.length) {
            obj.bridgeConfig = message.bridgeConfig.map((e) => exports.BridgeConfig.toJSON(e));
        }
        if (message.portalConfig?.length) {
            obj.portalConfig = message.portalConfig.map((e) => exports.PortalConfig.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.bridgeConfig = object.bridgeConfig?.map((e) => exports.BridgeConfig.fromPartial(e)) || [];
        message.portalConfig = object.portalConfig?.map((e) => exports.PortalConfig.fromPartial(e)) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
