// @remnawave/xtls-sdk/build/src/xray-protos/app/policy/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/policy/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config_LevelEntry = exports.Config = exports.SystemPolicy_Stats = exports.SystemPolicy = exports.Policy_Buffer = exports.Policy_Stats = exports.Policy_Timeout = exports.Policy = exports.Second = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.policy";
function createBaseSecond() {
    return { $type: "xray.app.policy.Second", value: 0 };
}
exports.Second = {
    $type: "xray.app.policy.Second",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.value !== 0) {
            writer.uint32(8).uint32(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSecond();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.value = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.Second.$type, value: isSet(object.value) ? globalThis.Number(object.value) : 0 };
    },
    toJSON(message) {
        const obj = {};
        if (message.value !== 0) {
            obj.value = Math.round(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.Second.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSecond();
        message.value = object.value ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Second.$type, exports.Second);
function createBasePolicy() {
    return { $type: "xray.app.policy.Policy", timeout: undefined, stats: undefined, buffer: undefined };
}
exports.Policy = {
    $type: "xray.app.policy.Policy",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.timeout !== undefined) {
            exports.Policy_Timeout.encode(message.timeout, writer.uint32(10).fork()).join();
        }
        if (message.stats !== undefined) {
            exports.Policy_Stats.encode(message.stats, writer.uint32(18).fork()).join();
        }
        if (message.buffer !== undefined) {
            exports.Policy_Buffer.encode(message.buffer, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePolicy();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.timeout = exports.Policy_Timeout.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.stats = exports.Policy_Stats.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.buffer = exports.Policy_Buffer.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Policy.$type,
            timeout: isSet(object.timeout) ? exports.Policy_Timeout.fromJSON(object.timeout) : undefined,
            stats: isSet(object.stats) ? exports.Policy_Stats.fromJSON(object.stats) : undefined,
            buffer: isSet(object.buffer) ? exports.Policy_Buffer.fromJSON(object.buffer) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.timeout !== undefined) {
            obj.timeout = exports.Policy_Timeout.toJSON(message.timeout);
        }
        if (message.stats !== undefined) {
            obj.stats = exports.Policy_Stats.toJSON(message.stats);
        }
        if (message.buffer !== undefined) {
            obj.buffer = exports.Policy_Buffer.toJSON(message.buffer);
        }
        return obj;
    },
    create(base) {
        return exports.Policy.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePolicy();
        message.timeout = (object.timeout !== undefined && object.timeout !== null)
            ? exports.Policy_Timeout.fromPartial(object.timeout)
            : undefined;
        message.stats = (object.stats !== undefined && object.stats !== null)
            ? exports.Policy_Stats.fromPartial(object.stats)
            : undefined;
        message.buffer = (object.buffer !== undefined && object.buffer !== null)
            ? exports.Policy_Buffer.fromPartial(object.buffer)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Policy.$type, exports.Policy);
function createBasePolicy_Timeout() {
    return {
        $type: "xray.app.policy.Policy.Timeout",
        handshake: undefined,
        connectionIdle: undefined,
        uplinkOnly: undefined,
        downlinkOnly: undefined,
    };
}
exports.Policy_Timeout = {
    $type: "xray.app.policy.Policy.Timeout",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.handshake !== undefined) {
            exports.Second.encode(message.handshake, writer.uint32(10).fork()).join();
        }
        if (message.connectionIdle !== undefined) {
            exports.Second.encode(message.connectionIdle, writer.uint32(18).fork()).join();
        }
        if (message.uplinkOnly !== undefined) {
            exports.Second.encode(message.uplinkOnly, writer.uint32(26).fork()).join();
        }
        if (message.downlinkOnly !== undefined) {
            exports.Second.encode(message.downlinkOnly, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePolicy_Timeout();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.handshake = exports.Second.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.connectionIdle = exports.Second.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.uplinkOnly = exports.Second.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.downlinkOnly = exports.Second.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Policy_Timeout.$type,
            handshake: isSet(object.handshake) ? exports.Second.fromJSON(object.handshake) : undefined,
            connectionIdle: isSet(object.connectionIdle) ? exports.Second.fromJSON(object.connectionIdle) : undefined,
            uplinkOnly: isSet(object.uplinkOnly) ? exports.Second.fromJSON(object.uplinkOnly) : undefined,
            downlinkOnly: isSet(object.downlinkOnly) ? exports.Second.fromJSON(object.downlinkOnly) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.handshake !== undefined) {
            obj.handshake = exports.Second.toJSON(message.handshake);
        }
        if (message.connectionIdle !== undefined) {
            obj.connectionIdle = exports.Second.toJSON(message.connectionIdle);
        }
        if (message.uplinkOnly !== undefined) {
            obj.uplinkOnly = exports.Second.toJSON(message.uplinkOnly);
        }
        if (message.downlinkOnly !== undefined) {
            obj.downlinkOnly = exports.Second.toJSON(message.downlinkOnly);
        }
        return obj;
    },
    create(base) {
        return exports.Policy_Timeout.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePolicy_Timeout();
        message.handshake = (object.handshake !== undefined && object.handshake !== null)
            ? exports.Second.fromPartial(object.handshake)
            : undefined;
        message.connectionIdle = (object.connectionIdle !== undefined && object.connectionIdle !== null)
            ? exports.Second.fromPartial(object.connectionIdle)
            : undefined;
        message.uplinkOnly = (object.uplinkOnly !== undefined && object.uplinkOnly !== null)
            ? exports.Second.fromPartial(object.uplinkOnly)
            : undefined;
        message.downlinkOnly = (object.downlinkOnly !== undefined && object.downlinkOnly !== null)
            ? exports.Second.fromPartial(object.downlinkOnly)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Policy_Timeout.$type, exports.Policy_Timeout);
function createBasePolicy_Stats() {
    return { $type: "xray.app.policy.Policy.Stats", userUplink: false, userDownlink: false, userOnline: false };
}
exports.Policy_Stats = {
    $type: "xray.app.policy.Policy.Stats",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.userUplink !== false) {
            writer.uint32(8).bool(message.userUplink);
        }
        if (message.userDownlink !== false) {
            writer.uint32(16).bool(message.userDownlink);
        }
        if (message.userOnline !== false) {
            writer.uint32(24).bool(message.userOnline);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePolicy_Stats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.userUplink = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.userDownlink = reader.bool();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.userOnline = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Policy_Stats.$type,
            userUplink: isSet(object.userUplink) ? globalThis.Boolean(object.userUplink) : false,
            userDownlink: isSet(object.userDownlink) ? globalThis.Boolean(object.userDownlink) : false,
            userOnline: isSet(object.userOnline) ? globalThis.Boolean(object.userOnline) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.userUplink !== false) {
            obj.userUplink = message.userUplink;
        }
        if (message.userDownlink !== false) {
            obj.userDownlink = message.userDownlink;
        }
        if (message.userOnline !== false) {
            obj.userOnline = message.userOnline;
        }
        return obj;
    },
    create(base) {
        return exports.Policy_Stats.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePolicy_Stats();
        message.userUplink = object.userUplink ?? false;
        message.userDownlink = object.userDownlink ?? false;
        message.userOnline = object.userOnline ?? false;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Policy_Stats.$type, exports.Policy_Stats);
function createBasePolicy_Buffer() {
    return { $type: "xray.app.policy.Policy.Buffer", connection: 0 };
}
exports.Policy_Buffer = {
    $type: "xray.app.policy.Policy.Buffer",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.connection !== 0) {
            writer.uint32(8).int32(message.connection);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePolicy_Buffer();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.connection = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Policy_Buffer.$type,
            connection: isSet(object.connection) ? globalThis.Number(object.connection) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.connection !== 0) {
            obj.connection = Math.round(message.connection);
        }
        return obj;
    },
    create(base) {
        return exports.Policy_Buffer.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBasePolicy_Buffer();
        message.connection = object.connection ?? 0;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Policy_Buffer.$type, exports.Policy_Buffer);
function createBaseSystemPolicy() {
    return { $type: "xray.app.policy.SystemPolicy", stats: undefined };
}
exports.SystemPolicy = {
    $type: "xray.app.policy.SystemPolicy",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.stats !== undefined) {
            exports.SystemPolicy_Stats.encode(message.stats, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSystemPolicy();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.stats = exports.SystemPolicy_Stats.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.SystemPolicy.$type,
            stats: isSet(object.stats) ? exports.SystemPolicy_Stats.fromJSON(object.stats) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.stats !== undefined) {
            obj.stats = exports.SystemPolicy_Stats.toJSON(message.stats);
        }
        return obj;
    },
    create(base) {
        return exports.SystemPolicy.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSystemPolicy();
        message.stats = (object.stats !== undefined && object.stats !== null)
            ? exports.SystemPolicy_Stats.fromPartial(object.stats)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.SystemPolicy.$type, exports.SystemPolicy);
function createBaseSystemPolicy_Stats() {
    return {
        $type: "xray.app.policy.SystemPolicy.Stats",
        inboundUplink: false,
        inboundDownlink: false,
        outboundUplink: false,
        outboundDownlink: false,
    };
}
exports.SystemPolicy_Stats = {
    $type: "xray.app.policy.SystemPolicy.Stats",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.inboundUplink !== false) {
            writer.uint32(8).bool(message.inboundUplink);
        }
        if (message.inboundDownlink !== false) {
            writer.uint32(16).bool(message.inboundDownlink);
        }
        if (message.outboundUplink !== false) {
            writer.uint32(24).bool(message.outboundUplink);
        }
        if (message.outboundDownlink !== false) {
            writer.uint32(32).bool(message.outboundDownlink);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSystemPolicy_Stats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.inboundUplink = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.inboundDownlink = reader.bool();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.outboundUplink = reader.bool();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.outboundDownlink = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.SystemPolicy_Stats.$type,
            inboundUplink: isSet(object.inboundUplink) ? globalThis.Boolean(object.inboundUplink) : false,
            inboundDownlink: isSet(object.inboundDownlink) ? globalThis.Boolean(object.inboundDownlink) : false,
            outboundUplink: isSet(object.outboundUplink) ? globalThis.Boolean(object.outboundUplink) : false,
            outboundDownlink: isSet(object.outboundDownlink) ? globalThis.Boolean(object.outboundDownlink) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.inboundUplink !== false) {
            obj.inboundUplink = message.inboundUplink;
        }
        if (message.inboundDownlink !== false) {
            obj.inboundDownlink = message.inboundDownlink;
        }
        if (message.outboundUplink !== false) {
            obj.outboundUplink = message.outboundUplink;
        }
        if (message.outboundDownlink !== false) {
            obj.outboundDownlink = message.outboundDownlink;
        }
        return obj;
    },
    create(base) {
        return exports.SystemPolicy_Stats.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseSystemPolicy_Stats();
        message.inboundUplink = object.inboundUplink ?? false;
        message.inboundDownlink = object.inboundDownlink ?? false;
        message.outboundUplink = object.outboundUplink ?? false;
        message.outboundDownlink = object.outboundDownlink ?? false;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.SystemPolicy_Stats.$type, exports.SystemPolicy_Stats);
function createBaseConfig() {
    return { $type: "xray.app.policy.Config", level: {}, system: undefined };
}
exports.Config = {
    $type: "xray.app.policy.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        Object.entries(message.level).forEach(([key, value]) => {
            exports.Config_LevelEntry.encode({ $type: "xray.app.policy.Config.LevelEntry", key: key, value }, writer.uint32(10).fork()).join();
        });
        if (message.system !== undefined) {
            exports.SystemPolicy.encode(message.system, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    const entry1 = exports.Config_LevelEntry.decode(reader, reader.uint32());
                    if (entry1.value !== undefined) {
                        message.level[entry1.key] = entry1.value;
                    }
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.system = exports.SystemPolicy.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            level: isObject(object.level)
                ? Object.entries(object.level).reduce((acc, [key, value]) => {
                    acc[globalThis.Number(key)] = exports.Policy.fromJSON(value);
                    return acc;
                }, {})
                : {},
            system: isSet(object.system) ? exports.SystemPolicy.fromJSON(object.system) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.level) {
            const entries = Object.entries(message.level);
            if (entries.length > 0) {
                obj.level = {};
                entries.forEach(([k, v]) => {
                    obj.level[k] = exports.Policy.toJSON(v);
                });
            }
        }
        if (message.system !== undefined) {
            obj.system = exports.SystemPolicy.toJSON(message.system);
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.level = Object.entries(object.level ?? {}).reduce((acc, [key, value]) => {
            if (value !== undefined) {
                acc[globalThis.Number(key)] = exports.Policy.fromPartial(value);
            }
            return acc;
        }, {});
        message.system = (object.system !== undefined && object.system !== null)
            ? exports.SystemPolicy.fromPartial(object.system)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function createBaseConfig_LevelEntry() {
    return { $type: "xray.app.policy.Config.LevelEntry", key: 0, value: undefined };
}
exports.Config_LevelEntry = {
    $type: "xray.app.policy.Config.LevelEntry",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.key !== 0) {
            writer.uint32(8).uint32(message.key);
        }
        if (message.value !== undefined) {
            exports.Policy.encode(message.value, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig_LevelEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.key = reader.uint32();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.value = exports.Policy.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config_LevelEntry.$type,
            key: isSet(object.key) ? globalThis.Number(object.key) : 0,
            value: isSet(object.value) ? exports.Policy.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== 0) {
            obj.key = Math.round(message.key);
        }
        if (message.value !== undefined) {
            obj.value = exports.Policy.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return exports.Config_LevelEntry.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig_LevelEntry();
        message.key = object.key ?? 0;
        message.value = (object.value !== undefined && object.value !== null)
            ? exports.Policy.fromPartial(object.value)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config_LevelEntry.$type, exports.Config_LevelEntry);
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
