// @remnawave/xtls-sdk/build/src/xray-protos/proxy/vless/account.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: proxy/vless/account.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Account = exports.Reverse = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.proxy.vless";
function createBaseReverse() {
    return { $type: "xray.proxy.vless.Reverse", tag: "" };
}
exports.Reverse = {
    $type: "xray.proxy.vless.Reverse",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tag !== "") {
            writer.uint32(10).string(message.tag);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseReverse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { $type: exports.Reverse.$type, tag: isSet(object.tag) ? globalThis.String(object.tag) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.tag !== "") {
            obj.tag = message.tag;
        }
        return obj;
    },
    create(base) {
        return exports.Reverse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseReverse();
        message.tag = object.tag ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Reverse.$type, exports.Reverse);
function createBaseAccount() {
    return {
        $type: "xray.proxy.vless.Account",
        id: "",
        flow: "",
        encryption: "",
        xorMode: 0,
        seconds: 0,
        padding: "",
        reverse: undefined,
        testpre: 0,
        testseed: [],
    };
}
exports.Account = {
    $type: "xray.proxy.vless.Account",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.flow !== "") {
            writer.uint32(18).string(message.flow);
        }
        if (message.encryption !== "") {
            writer.uint32(26).string(message.encryption);
        }
        if (message.xorMode !== 0) {
            writer.uint32(32).uint32(message.xorMode);
        }
        if (message.seconds !== 0) {
            writer.uint32(40).uint32(message.seconds);
        }
        if (message.padding !== "") {
            writer.uint32(50).string(message.padding);
        }
        if (message.reverse !== undefined) {
            exports.Reverse.encode(message.reverse, writer.uint32(58).fork()).join();
        }
        if (message.testpre !== 0) {
            writer.uint32(64).uint32(message.testpre);
        }
        writer.uint32(74).fork();
        for (const v of message.testseed) {
            writer.uint32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.flow = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.encryption = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.xorMode = reader.uint32();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.seconds = reader.uint32();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.padding = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.reverse = exports.Reverse.decode(reader, reader.uint32());
                    continue;
                }
                case 8: {
                    if (tag !== 64) {
                        break;
                    }
                    message.testpre = reader.uint32();
                    continue;
                }
                case 9: {
                    if (tag === 72) {
                        message.testseed.push(reader.uint32());
                        continue;
                    }
                    if (tag === 74) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.testseed.push(reader.uint32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Account.$type,
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            flow: isSet(object.flow) ? globalThis.String(object.flow) : "",
            encryption: isSet(object.encryption) ? globalThis.String(object.encryption) : "",
            xorMode: isSet(object.xorMode) ? globalThis.Number(object.xorMode) : 0,
            seconds: isSet(object.seconds) ? globalThis.Number(object.seconds) : 0,
            padding: isSet(object.padding) ? globalThis.String(object.padding) : "",
            reverse: isSet(object.reverse) ? exports.Reverse.fromJSON(object.reverse) : undefined,
            testpre: isSet(object.testpre) ? globalThis.Number(object.testpre) : 0,
            testseed: globalThis.Array.isArray(object?.testseed) ? object.testseed.map((e) => globalThis.Number(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.flow !== "") {
            obj.flow = message.flow;
        }
        if (message.encryption !== "") {
            obj.encryption = message.encryption;
        }
        if (message.xorMode !== 0) {
            obj.xorMode = Math.round(message.xorMode);
        }
        if (message.seconds !== 0) {
            obj.seconds = Math.round(message.seconds);
        }
        if (message.padding !== "") {
            obj.padding = message.padding;
        }
        if (message.reverse !== undefined) {
            obj.reverse = exports.Reverse.toJSON(message.reverse);
        }
        if (message.testpre !== 0) {
            obj.testpre = Math.round(message.testpre);
        }
        if (message.testseed?.length) {
            obj.testseed = message.testseed.map((e) => Math.round(e));
        }
        return obj;
    },
    create(base) {
        return exports.Account.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAccount();
        message.id = object.id ?? "";
        message.flow = object.flow ?? "";
        message.encryption = object.encryption ?? "";
        message.xorMode = object.xorMode ?? 0;
        message.seconds = object.seconds ?? 0;
        message.padding = object.padding ?? "";
        message.reverse = (object.reverse !== undefined && object.reverse !== null)
            ? exports.Reverse.fromPartial(object.reverse)
            : undefined;
        message.testpre = object.testpre ?? 0;
        message.testseed = object.testseed?.map((e) => e) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Account.$type, exports.Account);
function isSet(value) {
    return value !== null && value !== undefined;
}
