// @remnawave/xtls-sdk/build/src/xray-protos/app/log/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: app/log/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.LogType = exports.protobufPackage = void 0;
exports.logTypeFromJSON = logTypeFromJSON;
exports.logTypeToJSON = logTypeToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const log_1 = require("../../common/log/log");
const typeRegistry_1 = require("../../typeRegistry");
exports.protobufPackage = "xray.app.log";
var LogType;
(function (LogType) {
    LogType[LogType["None"] = 0] = "None";
    LogType[LogType["Console"] = 1] = "Console";
    LogType[LogType["File"] = 2] = "File";
    LogType[LogType["Event"] = 3] = "Event";
    LogType[LogType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(LogType || (exports.LogType = LogType = {}));
function logTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "None":
            return LogType.None;
        case 1:
        case "Console":
            return LogType.Console;
        case 2:
        case "File":
            return LogType.File;
        case 3:
        case "Event":
            return LogType.Event;
        case -1:
        case "UNRECOGNIZED":
        default:
            return LogType.UNRECOGNIZED;
    }
}
function logTypeToJSON(object) {
    switch (object) {
        case LogType.None:
            return "None";
        case LogType.Console:
            return "Console";
        case LogType.File:
            return "File";
        case LogType.Event:
            return "Event";
        case LogType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseConfig() {
    return {
        $type: "xray.app.log.Config",
        errorLogType: 0,
        errorLogLevel: 0,
        errorLogPath: "",
        accessLogType: 0,
        accessLogPath: "",
        enableDnsLog: false,
        maskAddress: "",
    };
}
exports.Config = {
    $type: "xray.app.log.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.errorLogType !== 0) {
            writer.uint32(8).int32(message.errorLogType);
        }
        if (message.errorLogLevel !== 0) {
            writer.uint32(16).int32(message.errorLogLevel);
        }
        if (message.errorLogPath !== "") {
            writer.uint32(26).string(message.errorLogPath);
        }
        if (message.accessLogType !== 0) {
            writer.uint32(32).int32(message.accessLogType);
        }
        if (message.accessLogPath !== "") {
            writer.uint32(42).string(message.accessLogPath);
        }
        if (message.enableDnsLog !== false) {
            writer.uint32(48).bool(message.enableDnsLog);
        }
        if (message.maskAddress !== "") {
            writer.uint32(58).string(message.maskAddress);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.errorLogType = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.errorLogLevel = reader.int32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.errorLogPath = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.accessLogType = reader.int32();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.accessLogPath = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.enableDnsLog = reader.bool();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.maskAddress = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            errorLogType: isSet(object.errorLogType) ? logTypeFromJSON(object.errorLogType) : 0,
            errorLogLevel: isSet(object.errorLogLevel) ? (0, log_1.severityFromJSON)(object.errorLogLevel) : 0,
            errorLogPath: isSet(object.errorLogPath) ? globalThis.String(object.errorLogPath) : "",
            accessLogType: isSet(object.accessLogType) ? logTypeFromJSON(object.accessLogType) : 0,
            accessLogPath: isSet(object.accessLogPath) ? globalThis.String(object.accessLogPath) : "",
            enableDnsLog: isSet(object.enableDnsLog) ? globalThis.Boolean(object.enableDnsLog) : false,
            maskAddress: isSet(object.maskAddress) ? globalThis.String(object.maskAddress) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.errorLogType !== 0) {
            obj.errorLogType = logTypeToJSON(message.errorLogType);
        }
        if (message.errorLogLevel !== 0) {
            obj.errorLogLevel = (0, log_1.severityToJSON)(message.errorLogLevel);
        }
        if (message.errorLogPath !== "") {
            obj.errorLogPath = message.errorLogPath;
        }
        if (message.accessLogType !== 0) {
            obj.accessLogType = logTypeToJSON(message.accessLogType);
        }
        if (message.accessLogPath !== "") {
            obj.accessLogPath = message.accessLogPath;
        }
        if (message.enableDnsLog !== false) {
            obj.enableDnsLog = message.enableDnsLog;
        }
        if (message.maskAddress !== "") {
            obj.maskAddress = message.maskAddress;
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.errorLogType = object.errorLogType ?? 0;
        message.errorLogLevel = object.errorLogLevel ?? 0;
        message.errorLogPath = object.errorLogPath ?? "";
        message.accessLogType = object.accessLogType ?? 0;
        message.accessLogPath = object.accessLogPath ?? "";
        message.enableDnsLog = object.enableDnsLog ?? false;
        message.maskAddress = object.maskAddress ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function isSet(value) {
    return value !== null && value !== undefined;
}
