// @remnawave/xtls-sdk/build/src/xray-protos/core/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.8.3
//   protoc               v5.29.3
// source: core/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutboundHandlerConfig = exports.InboundHandlerConfig = exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const typed_message_1 = require("../common/serial/typed_message");
const typeRegistry_1 = require("../typeRegistry");
exports.protobufPackage = "xray.core";
function createBaseConfig() {
    return { $type: "xray.core.Config", inbound: [], outbound: [], app: [], extension: [] };
}
exports.Config = {
    $type: "xray.core.Config",
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.inbound) {
            exports.InboundHandlerConfig.encode(v, writer.uint32(10).fork()).join();
        }
        for (const v of message.outbound) {
            exports.OutboundHandlerConfig.encode(v, writer.uint32(18).fork()).join();
        }
        for (const v of message.app) {
            typed_message_1.TypedMessage.encode(v, writer.uint32(34).fork()).join();
        }
        for (const v of message.extension) {
            typed_message_1.TypedMessage.encode(v, writer.uint32(50).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.inbound.push(exports.InboundHandlerConfig.decode(reader, reader.uint32()));
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.outbound.push(exports.OutboundHandlerConfig.decode(reader, reader.uint32()));
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.app.push(typed_message_1.TypedMessage.decode(reader, reader.uint32()));
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.extension.push(typed_message_1.TypedMessage.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Config.$type,
            inbound: globalThis.Array.isArray(object?.inbound)
                ? object.inbound.map((e) => exports.InboundHandlerConfig.fromJSON(e))
                : [],
            outbound: globalThis.Array.isArray(object?.outbound)
                ? object.outbound.map((e) => exports.OutboundHandlerConfig.fromJSON(e))
                : [],
            app: globalThis.Array.isArray(object?.app) ? object.app.map((e) => typed_message_1.TypedMessage.fromJSON(e)) : [],
            extension: globalThis.Array.isArray(object?.extension)
                ? object.extension.map((e) => typed_message_1.TypedMessage.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.inbound?.length) {
            obj.inbound = message.inbound.map((e) => exports.InboundHandlerConfig.toJSON(e));
        }
        if (message.outbound?.length) {
            obj.outbound = message.outbound.map((e) => exports.OutboundHandlerConfig.toJSON(e));
        }
        if (message.app?.length) {
            obj.app = message.app.map((e) => typed_message_1.TypedMessage.toJSON(e));
        }
        if (message.extension?.length) {
            obj.extension = message.extension.map((e) => typed_message_1.TypedMessage.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConfig();
        message.inbound = object.inbound?.map((e) => exports.InboundHandlerConfig.fromPartial(e)) || [];
        message.outbound = object.outbound?.map((e) => exports.OutboundHandlerConfig.fromPartial(e)) || [];
        message.app = object.app?.map((e) => typed_message_1.TypedMessage.fromPartial(e)) || [];
        message.extension = object.extension?.map((e) => typed_message_1.TypedMessage.fromPartial(e)) || [];
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Config.$type, exports.Config);
function createBaseInboundHandlerConfig() {
    return { $type: "xray.core.InboundHandlerConfig", tag: "", receiverSettings: undefined, proxySettings: undefined };
}
exports.InboundHandlerConfig = {
    $type: "xray.core.InboundHandlerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tag !== "") {
            writer.uint32(10).string(message.tag);
        }
        if (message.receiverSettings !== undefined) {
            typed_message_1.TypedMessage.encode(message.receiverSettings, writer.uint32(18).fork()).join();
        }
        if (message.proxySettings !== undefined) {
            typed_message_1.TypedMessage.encode(message.proxySettings, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInboundHandlerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.receiverSettings = typed_message_1.TypedMessage.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.proxySettings = typed_message_1.TypedMessage.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.InboundHandlerConfig.$type,
            tag: isSet(object.tag) ? globalThis.String(object.tag) : "",
            receiverSettings: isSet(object.receiverSettings) ? typed_message_1.TypedMessage.fromJSON(object.receiverSettings) : undefined,
            proxySettings: isSet(object.proxySettings) ? typed_message_1.TypedMessage.fromJSON(object.proxySettings) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.tag !== "") {
            obj.tag = message.tag;
        }
        if (message.receiverSettings !== undefined) {
            obj.receiverSettings = typed_message_1.TypedMessage.toJSON(message.receiverSettings);
        }
        if (message.proxySettings !== undefined) {
            obj.proxySettings = typed_message_1.TypedMessage.toJSON(message.proxySettings);
        }
        return obj;
    },
    create(base) {
        return exports.InboundHandlerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseInboundHandlerConfig();
        message.tag = object.tag ?? "";
        message.receiverSettings = (object.receiverSettings !== undefined && object.receiverSettings !== null)
            ? typed_message_1.TypedMessage.fromPartial(object.receiverSettings)
            : undefined;
        message.proxySettings = (object.proxySettings !== undefined && object.proxySettings !== null)
            ? typed_message_1.TypedMessage.fromPartial(object.proxySettings)
            : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.InboundHandlerConfig.$type, exports.InboundHandlerConfig);
function createBaseOutboundHandlerConfig() {
    return {
        $type: "xray.core.OutboundHandlerConfig",
        tag: "",
        senderSettings: undefined,
        proxySettings: undefined,
        expire: 0,
        comment: "",
    };
}
exports.OutboundHandlerConfig = {
    $type: "xray.core.OutboundHandlerConfig",
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.tag !== "") {
            writer.uint32(10).string(message.tag);
        }
        if (message.senderSettings !== undefined) {
            typed_message_1.TypedMessage.encode(message.senderSettings, writer.uint32(18).fork()).join();
        }
        if (message.proxySettings !== undefined) {
            typed_message_1.TypedMessage.encode(message.proxySettings, writer.uint32(26).fork()).join();
        }
        if (message.expire !== 0) {
            writer.uint32(32).int64(message.expire);
        }
        if (message.comment !== "") {
            writer.uint32(42).string(message.comment);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOutboundHandlerConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.tag = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.senderSettings = typed_message_1.TypedMessage.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.proxySettings = typed_message_1.TypedMessage.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.expire = longToNumber(reader.int64());
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.comment = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.OutboundHandlerConfig.$type,
            tag: isSet(object.tag) ? globalThis.String(object.tag) : "",
            senderSettings: isSet(object.senderSettings) ? typed_message_1.TypedMessage.fromJSON(object.senderSettings) : undefined,
            proxySettings: isSet(object.proxySettings) ? typed_message_1.TypedMessage.fromJSON(object.proxySettings) : undefined,
            expire: isSet(object.expire) ? globalThis.Number(object.expire) : 0,
            comment: isSet(object.comment) ? globalThis.String(object.comment) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.tag !== "") {
            obj.tag = message.tag;
        }
        if (message.senderSettings !== undefined) {
            obj.senderSettings = typed_message_1.TypedMessage.toJSON(message.senderSettings);
        }
        if (message.proxySettings !== undefined) {
            obj.proxySettings = typed_message_1.TypedMessage.toJSON(message.proxySettings);
        }
        if (message.expire !== 0) {
            obj.expire = Math.round(message.expire);
        }
        if (message.comment !== "") {
            obj.comment = message.comment;
        }
        return obj;
    },
    create(base) {
        return exports.OutboundHandlerConfig.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseOutboundHandlerConfig();
        message.tag = object.tag ?? "";
        message.senderSettings = (object.senderSettings !== undefined && object.senderSettings !== null)
            ? typed_message_1.TypedMessage.fromPartial(object.senderSettings)
            : undefined;
        message.proxySettings = (object.proxySettings !== undefined && object.proxySettings !== null)
            ? typed_message_1.TypedMessage.fromPartial(object.proxySettings)
            : undefined;
        message.expire = object.expire ?? 0;
        message.comment = object.comment ?? "";
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.OutboundHandlerConfig.$type, exports.OutboundHandlerConfig);
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
